### PR TITLE
feat: add typed interface

### PR DESCRIPTION
**Pull request by nilslice**
UPDATE: This has been tested to ensure it _does not_ break non-bundler JS programs. However, we don't provide any typings for those. Idk enough JS trickery to know if we could supply the type information to the globals (in Rust? as jsDoc?)

~UPDATE: I think this will break non-bundler JS programs though, since all the globals are renamed. Maybe I should re-work this so that maybe there's a container class/object that Http, Vars, Config, etc all live in vs. competing with the previous global names..~

The same import syntax stands, but I've reassigned the globals created in Rust to `__$NAME` variants in the Typescript, and then wrapped them in renamed classes to export from the library. 

---

If we publish this to a registry, it will enable something like: 

```ts
import { Host, Http, Var, Config } from '@extism/js-pdk'

export function greet() {
  Var.set("name", "MAYBESteve");
  let extra = new TextEncoder().encode("aaa")
  let decoded = new TextDecoder().decode(extra)
  const res = Http.request({ url: "https://example.com", method: "GET" });
  const name = Var.getString("name") || "unknown";
  const apiKey = Config.get("SOME_API_KEY") || "unknown";
  Host.outputString(`Hello, ${Host.inputString()} (or is it ${name}???) ${decoded} ${new Date().toString()}\n\n${res.body}\n\n ==== KEY: ${apiKey}`);
}
```

To build: 
```
make
cd examples/bundled
npm run build
extism call ../bundled.wasm greet --input "steve" --wasi --allow-host "*" --config SOME_API_KEY=123456789
```

### CODE REVIEW

File: crates/core/src/prelude/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
+import "core-js/actual/url";
+import "core-js/actual/url/to-json";
+import "core-js/actual/url-search-params";
+import { URLPattern } from "urlpattern-polyfill";
+
+declare module globalThis {
+  var URLPattern;
+  function __decodeUtf8BufferToString(...args): string;
+  function __encodeStringToUtf8Buffer(...args): ArrayBufferLike;
+  var __getTime;
```
Comment (bhelx) [inline crates/core/src/prelude/src/index.ts:10]: Do we need to expose these? I believe we actually delete these before we hand off control to the programmer. Is it possible to just add `// @ts-ignore` annotations to their use?

File: examples/host_funcs/script.d.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ declare module 'main' {
 
 declare module 'extism:host' {
   interface user {
-    myHostFunction1(ptr: I64): I64;
-    myHostFunction2(ptr: I64): I64;
+    myHostFunction1(ptr: PTR): PTR;
```
Comment (bhelx) [inline examples/host_funcs/script.d.ts:7]: 👍 nice catch

**Review by bhelx — APPROVED** [review]
Looks pretty good! We can follow up on pushing this out somewhere. we may be able to publish just the types to DefinitelyTyped instead of publishing a whole npm package out somewhere.

File: crates/core/src/prelude/src/index.ts (hunk 1 above)
Comment (nilslice) [reply crates/core/src/prelude/src/index.ts:10]: yea that's a good idea - I think I just got started trying to type everything before I knew what was actually needed in the globals

Comment (bhelx) [reply crates/core/src/prelude/src/index.ts:10]: These are kind of bootstrap functions. We write them in rust and expose to the JS layer, then we wrap them, then we delete their references. At least that's the idea. I can't recall if I actually delete all of them or not. we don't want people to use them directly though.

**Comment by nilslice** [comment]
I've added a `examples/kitchen-sink` plugin to run through a broader set of the PDK API. Noticed that the last chunk in that `greet` function is broken & unsure why. 

```ts
const mem = Memory.fromString(
    "Hello, " + body.data + " " + configLastName,
  );
```

This throws with a `not a function` error. But as far as I can tell, it is and calls through to the lower-level global `Memory` exposed from rust pdk.

File: crates/core/src/prelude/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# extism js-pdk types
+
+TypeScript definitions for the Extism [JS Plugin Development Kit](https://github.com/extism/js-pdk).
+
+To install these types, add them to your `tsconfig.json`'s `typeRoots`:
```
Comment (Phault) [inline crates/core/src/prelude/README.md:5]: While this works, it's not ideal as it doesn't set the appropriate default library, nor does it override the default of including all type packages in these folders, so you may end up with node, jest, etc globals.

I updated the root readme under the bundler section, with the usage I personally suggest.

Comment (chrisdickinson) [reply crates/core/src/prelude/README.md]: Ah, good catch, thank you!

**Pull request merged**